tools/server/utils.hpp

```diff
@@ -437,8 +437,11 @@ template <class Iter>
 static std::string tokens_to_str(llama_context * ctx, Iter begin, Iter end) {
     std::string ret;
     for (; begin != end; ++begin) {
-        ret += common_token_to_piece(ctx, *begin);
+        std::string piece = common_token_to_piece(ctx, *begin);
+        fprintf(stderr, "[DEBUG] Token ID: %d -> Piece: '%s' (length: %zu)\n", *begin, piece.c_str(), piece.length());
```

Comment:
Same comment about logging for these ones. They're definitely useful, so it would be great to make these proper logs!